### ISSUE TITLE
how can I exec code (python/R) by kernel gateway?

### ISSUE THREAD

**Issue by 15050050972** [jupyter/kernel_gateway#312]
I didn't find corresponding interface in swagger.yaml

**Comment by rolweber** [comment]
Kernel Gateway provides the same REST interfaces for managing kernels and sessions as Jupyter Notebook does. We don't have separate documentation for that.

**Comment by rolweber** [comment]
If you're using JavaScript, try the [`@jupyterlab/services`](https://www.npmjs.com/package/@jupyterlab/services) node module.
There's a link to REST API docs there as well:
http://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyter/notebook/master/notebook/services/api/api.yaml

The "contents" REST APIs are not available in Kernel Gateway. I'm not sure about terminals or config.

**Comment by 15050050972** [comment]
It seems I can't exec code by REST API.

But I found I can exec code by using websocket : https://jupyter-client.readthedocs.io/en/latest/messaging.html#execute

I start the jupyter gateway, i test it by REST API, and i get correct response. see: http://10.30.32.6:8888/api/kernelspecs

But when I try websocket, I get an error 
![image](https://user-images.githubusercontent.com/23148429/51664142-b07ce580-1ff3-11e9-99af-73cefbce5d5e.png)
Do I input correct websocket url?
 I read the offical document above, but I still have no idea to solve it. And I didn't find any simple examples by using websocket with javascirpt in jupyter or jupyter-kernel

**Comment by rolweber** [comment]
Some excerpts from my JS client...

A handler for printing messages from the kernel, including stdout and execution results:
```javascript
function iopubHandler(kernel, message) {

    switch (message.msg_type) {

    case "status":
        console.log("### STATUS", message.content);
        break;

    case "stream":
        console.log("### OUTPUT", message.content.name);
        console.log(message.content.text);
        console.log("### end of output");
        break;

    case "execute_input":
        console.log("### EXECUTE", message.content.execution_count);
        console.log(message.content.code);
        console.log("### end of code");
        break;

    case "execute_result":
        console.log("### RESULT #", message.content.execution_count);
        console.log(message.content.data);
        console.log("### end of result");
        break;

    case "display_data": // R kernel sends JSON result
        if (message.content.data && message.content.data["text/plain"]) {
            console.log("### DISPLAY", "text/plain");
            console.log(message.content.data["text/plain"]);
            console.log("### end of display data");
        } else {
            console.log("### DISPLAY", message.content);
            console.log("### end of display data");
        } break;

    case "error":
        console.log("### ERROR", message.content);
        console.log("### end of error");
        break;

    default:
        console.log("### iopubHandler", kernel._id, message.msg_type);
        break;
    }
}
```

A function to register the handler and send a series of code snippets for execution to a kernel:
```javascript
function executeSnippets(kernel, snippets) {
    kernel.iopubMessage.connect(iopubHandler);

    var i = 0;

    var nextCode = () => {
        if (i < snippets.length) {
            var req = { code: snippets[i++] };
            kernel.requestExecute(req).onReply = nextCode;
        }
    };
    nextCode();
}
```

**Comment by 15050050972** [comment]
Thanks very much! 
But Why I can't connect the jupyter kernel by websocket ? If I can't connect success, I think i can't implements any function

**Comment by 15050050972** [comment]
I try to start a kernel by using REST API, but I got an 404 error ? 
![image](https://user-images.githubusercontent.com/23148429/51666811-7dd5eb80-1ff9-11e9-82b3-0e548da93167.png)
I followed the official document : https://github.com/jupyter/kernel_gateway/blob/master/kernel_gateway/jupyter_websocket/swagger.yaml

**Comment by rolweber** [comment]
You might be missing a path prefix. Check the `base_url` setting in the config.
https://jupyter-notebook.readthedocs.io/en/stable/config.html
Kernel Gateway logs a URL at startup, but that URL is missing the path.

Also, for the first request, I'd rather fetch the kernelspecs instead of starting a kernel.

**Comment by kevin-bates** [comment]
You might find this [gateway_client](https://github.com/jupyter/enterprise_gateway/blob/master/enterprise_gateway/client/gateway_client.py) helpful if you need python.

**Comment by 15050050972** [comment]
I can open a websocket connection, but why I can't get any response ?
here is my code, thanks
```
<script>
        var session = Date.now().toString();
        var msg = {
            header: {
                msg_id: Date.now().toString(),
                username: 'test',
                session: session,
                msg_type: 'execute_request',
                version: '5.0'
            },
            // parent_header: 'dict',
            // metadata: 'dict',
            content: {
                code : 'print(11)',
                allow_stdin: false
            },
            // buffers: 'list'
        }
        var str = "/b0a7cd45-3c97-4037-985c-45464e99cfa1";
        var ws = new WebSocket("ws://10.30.32.6:8888/api/kernels/b0a7cd45-3c97-4037-985c-45464e99cfa1/channels");
        ws.onopen = function () {
            // ws.send("发送数据");
            console.log('open', msg);
            ws.send(msg);
        };

        ws.onmessage = function (evt) {
            var received_msg = evt.data;
            console.log('recv:', received_msg);
        };

        ws.onclose = function () {
            // 关闭 websocket
            console.log('open');
        };
        // console.log(ws);
    </script>
```

**Comment by 15050050972** [comment]
I try this, and I did get an error...
```
[KernelGatewayApp] Initializing websocket connection /api/kernels/d3b65a99-6b75-419b-a3e7-4c88310f1a22/channels
[KernelGatewayApp] WARNING | No session ID specified
[KernelGatewayApp] Adapting to protocol v5.1 for kernel d3b65a99-6b75-419b-a3e7-4c88310f1a22
[I 190125 14:11:51 web:2162] 101 GET /api/kernels/d3b65a99-6b75-419b-a3e7-4c88310f1a22/channels (10.30.32.2) 2.57ms
[KernelGatewayApp] Opening websocket /api/kernels/d3b65a99-6b75-419b-a3e7-4c88310f1a22/channels
[KernelGatewayApp] Getting buffer for d3b65a99-6b75-419b-a3e7-4c88310f1a22
[KernelGatewayApp] Clearing buffer for d3b65a99-6b75-419b-a3e7-4c88310f1a22
[KernelGatewayApp] Connecting to: tcp://127.0.0.1:38831
[KernelGatewayApp] Connecting to: tcp://127.0.0.1:45140
[KernelGatewayApp] Connecting to: tcp://127.0.0.1:38686
[E 190125 14:11:51 web:1670] Uncaught exception GET /api/kernels/d3b65a99-6b75-419b-a3e7-4c88310f1a22/channels (10.30.32.2)
    HTTPServerRequest(protocol='http', host='10.30.32.6:8888', method='GET', uri='/api/kernels/d3b65a99-6b75-419b-a3e7-4c88310f1a22/channels', version='HTTP/1.1', remote_ip='10.30.32.2')
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/site-packages/tornado/websocket.py", line 546, in _run_callback
        result = callback(*args, **kwargs)
      File "/usr/local/lib/python3.6/site-packages/notebook/services/kernels/handlers.py", line 301, in on_message
        msg = json.loads(msg)
      File "/usr/local/lib/python3.6/json/__init__.py", line 354, in loads
        return _default_decoder.decode(s)
      File "/usr/local/lib/python3.6/json/decoder.py", line 339, in decode
        obj, end = self.raw_decode(s, idx=_w(s, 0).end())
      File "/usr/local/lib/python3.6/json/decoder.py", line 357, in raw_decode
        raise JSONDecodeError("Expecting value", s, err.value) from None
    json.decoder.JSONDecodeError: Expecting value: line 1 column 2 (char 1)
[KernelGatewayApp] Websocket closed d3b65a99-6b75-419b-a3e7-4c88310f1a22:a243021c-405c19b7fcf45501e0240a96
[KernelGatewayApp] Starting buffering for d3b65a99-6b75-419b-a3e7-4c88310f1a22:a243021c-405c19b7fcf45501e0240a96
[KernelGatewayApp] Clearing buffer for d3b65a99-6b75-419b-a3e7-4c88310f1a22
```
Code is still above(kernelId is new)
It seems like json parse failure. But I follw the offical document (https://jupyter-client.readthedocs.io/en/latest/messaging.html#general-message-format)

**Comment by 15050050972** [comment]
Yes, I am try to re-implement the Jupyter Messaging protocal, thanks for your time. I will look at the @jupyterlab/service source code.
